pg_search/src/postgres/storage/fsm.rs

```diff
@@ -34,413 +35,1263 @@ enum FSMBlockKind {
+                    if next_blockno == pg_sys::InvalidBlockNumber && cnt == 0 {
+                        let mut root = bman.get_buffer_mut(self.start_blockno);
+                        let mut page = root.page_mut();
+                        let mut tree = self.avl_mut(&mut page);
+
+                        // ok if another backend already removed it.
+                        let _ = tree.remove(&found_xid);
+                        drop(root);
+
+                        // move to the next candidate XID below this one.
+                        xid = found_xid - 1;
+                        if xid < pg_sys::FirstNormalTransactionId.into_inner() as u64 {
+                            break;
+                        }
+                        continue 'outer;
+                    }
+
+                    // advance to the next block in the chain.
+                    blockno = next_blockno;
                 }
 
-                if all_invalid {
-                    // the page is now all invalid so mark it as empty
-                    contents.header.empty = true;
-                } else if current_block_count == blocks.len() {
-                    // we didn't actually change the page
-                    buffer.set_dirty(false);
-                    continue;
+                if blocks.len() == many {
+                    break;
                 }
 
-                if blocks.len() == n {
-                    // we have all the requested blocks so return them
-                    return blocks.into_iter();
+                // exhausted this list
+                // move to the next candidate XID below this one.
+                xid = found_xid - 1;
+                if xid < pg_sys::FirstNormalTransactionId.into_inner() as u64 {
+                    break;
                 }
-            } else {
-                // page was empty by the time we got the exclusive lock -- we didn't change it
-                buffer.set_dirty(false);
             }
+
+            blocks.into_iter()
         }
-    }
 
-    /// Add the specified `extend_with` iterator of [`pg_sys::BlockNumber`]s to this [`FreeSpaceManager`].
-    ///
-    /// The added blocks will be recyclable in the future based on the current [`pg_sys::GetCurrentTransactionId`].
-    pub fn extend(
-        &self,
-        bman: &mut BufferManager,
-        extend_with: impl Iterator<Item = pg_sys::BlockNumber>,
-    ) {
-        let current_xid =
-            unsafe { pg_sys::GetCurrentTransactionIdIfAny().max(pg_sys::FirstNormalTransactionId) };
-        self.extend_with_when_recyclable(bman, current_xid, extend_with);
-    }
+        fn extend_with_when_recyclable(
+            &mut self,
+            bman: &mut BufferManager,
+            when_recyclable: pg_sys::FullTransactionId,
+            extend_with: impl Iterator<Item = pg_sys::BlockNumber>,
+        ) -> bool {
+            let tag = {
```

Comment:
We're not holding any locks after acquiring this tag. Is it possible for a `drain` call to remove it before we are able to use it, or is that impossible due to the properties of the transaction ids?

Maybe worth a comment.

Reply:
yeah, that's actually a flaw.  fixed in 77c1d615a6cc0cea85eab7faa271763c9ee939fb